server/events.go

```diff
@@ -901,6 +915,10 @@ func (s *Server) accConnsUpdate(a *Account) {
 	s.sendAccConnsUpdate(a, subj)
 }
 
+func (s *Server) nextEventID() string {
+	return s.sys.eventids.Next()
```

Comment:
@kozlovic is this enough here or do I need to check for nil on `sys` and fall back to the global `nuid`?

Reply:
```
WARNING: DATA RACE

Read at 0x00c0002934e8 by goroutine 94:

  github.com/nats-io/nats-server/vendor/github.com/nats-io/nuid.(*NUID).Next()

      /home/travis/gopath/src/github.com/nats-io/nats-server/vendor/github.com/nats-io/nuid/nuid.go:96 +0x5b

  github.com/nats-io/nats-server/server.(*Server).nextEventID()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/events.go:919 +0x21b

  github.com/nats-io/nats-server/server.(*Server).accountDisconnectEvent()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/events.go:982 +0x3fa

  github.com/nats-io/nats-server/server.(*Server).saveClosedClient()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/server.go:1924 +0x169

Previous write at 0x00c0002934e8 by goroutine 369:

  github.com/nats-io/nats-server/vendor/github.com/nats-io/nuid.(*NUID).Next()

      /home/travis/gopath/src/github.com/nats-io/nats-server/vendor/github.com/nats-io/nuid/nuid.go:96 +0x9d

  github.com/nats-io/nats-server/server.(*Server).nextEventID()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/events.go:919 +0x21b

  github.com/nats-io/nats-server/server.(*Server).accountDisconnectEvent()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/events.go:982 +0x3fa

  github.com/nats-io/nats-server/server.(*Server).saveClosedClient()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/server.go:1924 +0x169

Goroutine 94 (running) created at:

  github.com/nats-io/nats-server/server.(*client).markConnAsClosed()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/client.go:1214 +0x1dc

  github.com/nats-io/nats-server/server.(*client).flushOutbound()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/client.go:1093 +0x146a

  github.com/nats-io/nats-server/server.(*client).writeLoop()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/client.go:811 +0x1fe

  github.com/nats-io/nats-server/server.(*Server).createLeafNode.func4()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/leafnode.go:800 +0x41

Goroutine 369 (finished) created at:

  github.com/nats-io/nats-server/server.(*client).markConnAsClosed()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/client.go:1214 +0x1dc

  github.com/nats-io/nats-server/server.(*client).flushOutbound()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/client.go:1093 +0x146a

  github.com/nats-io/nats-server/server.(*client).writeLoop()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/client.go:811 +0x1fe

  github.com/nats-io/nats-server/server.(*Server).createLeafNode.func4()

      /home/travis/gopath/src/github.com/nats-io/nats-server/server/leafnode.go:800 +0x41

==================

    TestLeafNodeConnectionLimitsCluster: testing.go:906: race detected during execution of test

--- FAIL: TestLeafNodeConnectionLimitsCluster (0.46s)

    : testing.go:906: race detected during execution of test
``` 

Somethings screwy with what I did, but it seems I did the right thing unless this somehow relates to a late initialization

Reply:
You have to protect with a lock when you create your own. Nuids are not protected by default, just the global one.

Reply:
The idea was to store an instance and get the next UID from a place where we already had the lock. For instance, in the 2 places right now where we use that, we check if events are enabled under the server lock. So it could have been captured there and then used after releasing the server lock, for instance in your PR that would be line 931, 970 and 1018.

Reply:
For instance:
```
        s.mu.Lock()
	if !s.eventsEnabled() {
		s.mu.Unlock()
		return
	}
        id := s.nextEventID()
	s.mu.Unlock()
	now := time.Now()
	c.mu.Lock()
	m := DisconnectEventMsg{
		Type: DisconnectEventMsgType,
		ID:   id,
...
```